test/uk/gov/hmrc/cgtpropertydisposals/connectors/ConnectorSpec.scala

```diff
@@ -0,0 +1,216 @@
+import org.scalatest.{BeforeAndAfterAll, BeforeAndAfterEach, EitherValues}
+import org.scalatestplus.play.guice.GuiceOneServerPerSuite
+import play.api.http.{HeaderNames, Status}
+import play.api.inject.guice.GuiceApplicationBuilder
+import play.api.libs.json.{JsObject, JsValue, Json}
+import play.api.libs.ws.WSClient
+import play.api.test.{DefaultAwaitTimeout, FutureAwaits, ResultExtractors}
+import play.api.{Application, Environment, Mode}
+import play.shaded.ahc.org.asynchttpclient.exception.RemotelyClosedException
+import support.{WireMockHelper, WireMockMethods}
+import uk.gov.hmrc.cgtpropertydisposals.connectors.ConnectorSpec.{onDownstreamFault, onDownstreamResponse}
+import uk.gov.hmrc.cgtpropertydisposals.models.Error
+import uk.gov.hmrc.http.{GatewayTimeoutException, HeaderCarrier, HttpResponse}
+
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+
+trait ConnectorSpec
+    extends AnyWordSpecLike
+    with EitherValues
+    with Matchers
+    with FutureAwaits
+    with DefaultAwaitTimeout
+    with WireMockHelper
+    with GuiceOneServerPerSuite
+    with BeforeAndAfterEach
+    with BeforeAndAfterAll
+    with Status
+    with HeaderNames
+    with ResultExtractors {
+
+  implicit val hc: HeaderCarrier = HeaderCarrier()
+
+  val mockHost: String = WireMockHelper.host
+  val mockPort: String = WireMockHelper.wireMockPort.toString
+
+  lazy val client: WSClient = app.injector.instanceOf[WSClient]
+
+  val requiredHosts: Seq[String]
+
+  def servicesConfig: Map[String, String] = requiredHosts.foldLeft(Map.empty[String, String])((config, host) =>
+    config ++
+      Seq(
+        s"microservice.services.$host.host" -> mockHost,
+        s"microservice.services.$host.port" -> mockPort
+      )
+  )
+
+  override implicit lazy val app: Application = new GuiceApplicationBuilder()
+      .in(Environment.simple(mode = Mode.Dev))
+      .configure(servicesConfig)
+      .build()
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    startWireMock()
+  }
+
+  override def afterAll(): Unit = {
+    stopWireMock()
+    super.afterAll()
+  }
+
+  override def beforeEach(): Unit = {
+    resetWireMock()
+    super.beforeEach()
+  }
+
+  def singleDownstreamErrorJson(errorCode: String, errorReason: String): JsValue = Json.parse(
+    s"""
+      |{
+      |   "code": "$errorCode",
+      |   "reason": "$errorReason"
+      |}
+    """.stripMargin
+  )
+
+  case class TestScenario(scenarioName: String,
+                          returnStatus: Int,
+                          expectedStatus: Int,
+                          returnBody: JsValue,
+                          expectedBody: JsValue,
+                          isErrorScenario: Boolean)
+
+  def testGET(downstreamUrl: String,
+              downstreamQueryParams: Map[String, String],
+              testScenarios: Seq[TestScenario],
+              connectorAction: () => EitherT[Future, Error, HttpResponse]): Unit = {
```

Comment:
I was considering building something like this, but ultimately decided against it. The abstraction layer hinders debugging and diagnosing test failures. And it's already pretty difficult to debug these things as is. I think a little bit of repetition in these tests is probably OK.